design/20221205-memory-management.md

```diff
@@ -0,0 +1,682 @@
+
+Here is a script that sets up the issuers, creates the `Certificate`s, waits for them to become ready and outputs the total time taken https://gist.github.com/irbekrm/bc56a917a164b1a3a097bda483def0b8.
+
+##### latest cert-manager
+
+This test was run against a version of cert-manager that corresponds to v1.11.0-alpha.2 with some added client-go metrics https://github.com/irbekrm/cert-manager/tree/client_go_metrics.
+Run a script to set up 10 CA issuers, create 500 certificates and observe the time taken for all certs to be issued:
+![alt text](https://github.com/irbekrm/cert-manager/blob/memory_design/design/images/20221205-memory-management/masterissuanceterminal.png?raw=true)
+
+Observe resource consumption, request rate and latency for cert-manager controller:
+![alt text](https://github.com/irbekrm/cert-manager/blob/memory_design/design/images/20221205-memory-management/mastercertmanager.png?raw=true)
+
+Observe resource consumption and rate of requests for `Secret` resources for kube apiserver:
+![alt text](https://github.com/irbekrm/cert-manager/blob/memory_design/design/images/20221205-memory-management/masterkubeapiserver.png?raw=true)
+
+##### partial metadata
+
+Run a script to set up 10 CA issuers, create 500 certificates and observe the time taken for all certs to be issued:
+![alt text](https://github.com/irbekrm/cert-manager/blob/memory_design/design/images/20221205-memory-management/partialnolabels.png?raw=true)
+
+Observe resource consumption, request rate and latency for cert-manager controller:
+![alt text](https://github.com/irbekrm/cert-manager/blob/memory_design/design/images/20221205-memory-management/partialnolabelscertmanager.png?raw=true)
+
+Observe resource consumption and rate of requests for `Secret` resources for kube apiserver:
+![alt text](https://github.com/irbekrm/cert-manager/blob/memory_design/design/images/20221205-memory-management/partialnolabelskubeapiserver.png?raw=true)
+
+The issuance is slightly slowed down because on each issuance cert-manager needs to get the unlabelled CA `Secret` directly from kube apiserver.
+Users could mitigate this by adding cert-manager labels to the CA `Secret`s.
+Run a modified version of the same script, but [with CA `Secret`s labelled](https://gist.github.com/irbekrm/bc56a917a164b1a3a097bda483def0b8#file-measure-issuance-time-sh-L31-L34):
+
+![alt text](https://github.com/irbekrm/cert-manager/blob/memory_design/design/images/20221205-memory-management/partiallabels.png?raw=true)
+
+### Pros
+
+- In most setups in majority of cases where a control loop needs a `Secret` it would still be retrieved from cache (as it is certificate secrets that get parsed most frequently and those will be labelled in practically all cases)
+
+- Memory consumption improvements appear quite significant
+
+- Once graduated to GA would work for all installations without needing to discover a flag to set
+
+### Cons
+
+- All cluster `Secret`s are still listed
```

Comment:
So I am a bit concerned that this exposes a mechanism for end-users of cert-manager's APIs to intentionally or unintentionally cause cert-manager's controller to create an excess number of requests to the kubernetes apiserver (which could either degrade apiserver performance, _or_ act as an effective DoS for other clients as our clientset could get ratelimited, either client-side, SSA or any other means).

One case I could see where this is unintentional is say, a flapping GitOps system.

If I am running the controller as a managed offering for many tenants, how can I protect myself and other users from this?

Reply:
I've looked at this yeah https://github.com/cert-manager/cert-manager/blob/65be2caaaeadbfb8dff63c1afeb01522cff52805/design/20221205-memory-management.md#issuance-of-a-large-number-of-certificates here are the metrics

In practice it would be one additional GET call per issuance for issuers that are set up with some secret (which is all of them except for self-signed I think) and one additional per issuer setup.
In regards to unintentionally overwhelming the API server, I think the metrics show that the additional number of calls and issuance time increase is acceptable. Users could then improve it by labelling issuer secrets. And given that this would start as an alpha feature we should be able to gather feedback from larger setups before this becomes enabled by default.
In regards to intentionally overwhelming, again I think one GET call per issuance and per issuer setup does not make this particular feature that much more possible to use cert-manager for this? I don't know if there is much more we can do in general to prevent this - as I understand the way to prevent this kind of problem is API Priority and Fairness server-side

Reply:
I've added overwhelming kube apiserver to risks and mitigations section https://github.com/cert-manager/cert-manager/blob/47dcce7c70f2e33be78f5e9d56d22ea887b60370/design/20221205-memory-management.md#risks-and-mitigations
I've also read through the issuers code to check how often the secrets get retrieved and updated the issuer secrets section https://github.com/cert-manager/cert-manager/blob/47c3c4c5f4e5effbb32a5f156efe2140fa67961a/design/20221205-memory-management.md#secrets-for-clusterissuers

> One case I could see where this is unintentional is say, a flapping GitOps system.

The system then would need to make actual changes to either issuer config or issuer secrets or certificates for this to happen. With such changes to certificates (no AP&F and cert-manager configured with very high QPS and burst) we could already overwhelm apiserver today (as we'd be issuing in a tight loop). I think this is unavoidable and would be up to users to limit who can modify cert-manager installation and to have AP&F